// @livepeer/react/dist/useTheme-77973284.cjs.dev.js
'use strict';

var context = require('@livepeer/core-react/context');
var media = require('livepeer/media');
var browser = require('livepeer/media/browser');
var React = require('react');
var create = require('zustand');
var jsxRuntime = require('react/jsx-runtime');
var livepeer = require('livepeer');
var styling = require('livepeer/media/browser/styling');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n["default"] = e;
  return Object.freeze(n);
}

var React__namespace = /*#__PURE__*/_interopNamespace(React);
var create__default = /*#__PURE__*/_interopDefault(create);

var ThemeContext = /*#__PURE__*/React__namespace.createContext(null);

var MediaControllerContext = /*#__PURE__*/React__namespace.createContext(create__default["default"](media.createControllerStore({
  element: null,
  device: browser.getDeviceInfo(),
  storage: livepeer.createStorage({}),
  opts: {}
})));

var MediaControllerProvider = function MediaControllerProvider(_ref) {
  var element = _ref.element,
    children = _ref.children,
    opts = _ref.opts;
  var useMediaController = useMediaControllerStore(element, opts);
  return /*#__PURE__*/jsxRuntime.jsx(MediaControllerContext.Provider, {
    value: useMediaController,
    children: children
  });
};
var useMediaControllerStore = function useMediaControllerStore(element, opts) {
  var client = context.useClient();
  var store = React__namespace.useMemo(function () {
    return create__default["default"](media.createControllerStore({
      element: element !== null && element !== void 0 ? element : null,
      device: browser.getDeviceInfo(),
      storage: client.storage,
      opts: opts !== null && opts !== void 0 ? opts : {}
    }));
  }, [element, client === null || client === void 0 ? void 0 : client.storage, opts]);
  React__namespace.useEffect(function () {
    var _addEventListeners = browser.addEventListeners(store, opts),
      destroy = _addEventListeners.destroy;
    return function () {
      destroy === null || destroy === void 0 ? void 0 : destroy();
    };
  }, [store, opts]);
  return store;
};

var useMediaController = function useMediaController(selector) {
  var useMediaControllerInternal = React__namespace.useContext(MediaControllerContext);
  return useMediaControllerInternal(selector);
};

var useTheme = function useTheme(theme) {
  return context.useTheme(ThemeContext, styling.createTheme, theme);
};

exports.MediaControllerContext = MediaControllerContext;
exports.MediaControllerProvider = MediaControllerProvider;
exports.ThemeContext = ThemeContext;
exports.useMediaController = useMediaController;
exports.useTheme = useTheme;
